src/mlpack/methods/reinforcement_learning/ppo_impl.hpp

```diff
@@ -0,0 +1,276 @@
+
+  if (criticNetwork.Parameters().is_empty())
+    criticNetwork.ResetParameters();
+
+  this->criticUpdater.Initialize(criticNetwork.Parameters().n_rows,
+      criticNetwork.Parameters().n_cols);
+
+  this->actorUpdater.Initialize(actorNetwork.Parameters().n_rows,
+      actorNetwork.Parameters().n_cols);
+  oldActorNetwork = actorNetwork;
+}
+
+template<
+  typename EnvironmentType,
+  typename ActorNetworkType,
+  typename CriticNetworkType,
+  typename UpdaterType,
+  typename PolicyType,
+  typename ReplayType
+>
+void PPO<
+  EnvironmentType,
+  ActorNetworkType,
+  CriticNetworkType,
+  UpdaterType,
+  PolicyType,
+  ReplayType
+>::Update()
+{
+  // Sample from previous experience.
+  arma::mat sampledStates;
+  std::vector<ActionType> sampledActions;
+  arma::colvec sampledRewards;
+  arma::mat sampledNextStates;
+  arma::icolvec isTerminal;
+
+  replayMethod.Sample(sampledStates, sampledActions, sampledRewards,
+      sampledNextStates, isTerminal);
+
+  arma::rowvec discountedRewards(sampledRewards.n_rows);
+  arma::mat nextActionValues;
+  double values = 0.0;
+  criticNetwork.Predict(sampledNextStates, nextActionValues);
+  for (size_t i = sampledRewards.n_cols; i > 0; --i)
+  {
+    values = sampledRewards[i - 1] + values * config.Discount();
+    discountedRewards[sampledRewards.n_cols - i] = values;
+  }
+
+  arma::mat actionValues, advantages, criticGradients, actorGradients;
+  criticNetwork.Forward(sampledStates, actionValues);
+
+  advantages = arma::conv_to<arma::mat>::
+      from(discountedRewards) - actionValues;
+
+  // Update the critic.
+  criticNetwork.Backward(advantages, criticGradients);
+  criticUpdater.Update(criticNetwork.Parameters(), config.StepSize(),
+      criticGradients);
+
+  // calculate the ratio.
+  arma::mat actionParameter, sigma, mu;
+  actorNetwork.Forward(sampledStates, actionParameter);
+
+  ann::TanhFunction::Fn(actionParameter.row(0), mu);
+  ann::SoftplusFunction::Fn(actionParameter.row(1), sigma);
+
+  ann::NormalDistribution normalDist =
+      ann::NormalDistribution(vectorise(mu, 0), vectorise(sigma, 0));
+
+  oldActorNetwork.Forward(sampledStates, actionParameter);
+  ann::TanhFunction::Fn(actionParameter.row(0), mu);
+  ann::SoftplusFunction::Fn(actionParameter.row(1), sigma);
+
+  ann::NormalDistribution oldNormalDist =
+      ann::NormalDistribution(vectorise(mu, 0), vectorise(sigma, 0));
+
+  // Update the actor.
+  // observation use action.
+  arma::vec prob, oldProb;
+  arma::colvec observation(sampledActions.size());
+  for (size_t i = 0; i < sampledActions.size(); i ++)
+  {
+    observation[i] = sampledActions[i].action;
+  }
+  normalDist.Probability(observation, prob);
+  oldNormalDist.Probability(observation, oldProb);
+
+  arma::mat ratio = vectorise((prob / oldProb), 1);
+
+  arma::mat surrogateLoss = arma::clamp(ratio, 1 - config.Epsilon(),
+      1 + config.Epsilon()) % advantages;
+  arma::mat loss = - arma::min(ratio % advantages, surrogateLoss);
```

Comment:
Not sure I get this, where is this coming from?

Reply:
Don't get me wrong there is nothing wrong with the line, I just don't get where this is coming from.

Reply:
> Don't get me wrong there is nothing wrong with the line, I just don't get where this is coming from.

Do you mean the equation of loss calculation? It is the PPO-Clip method. You can refer to [there](https://spinningup.openai.com/en/latest/algorithms/ppo.html#key-equations) or in this [paper](https://arxiv.org/pdf/1707.06347.pdf).

Reply:
Exactly, thanks for the reference.